src/ring_theory/algebra.lean

```diff
@@ -26,55 +26,37 @@ ring is the under category R ↓ CRing. In the categorical
 setting we have a forgetful functor R-Alg ⥤ R-Mod.
 However here it extends module in order to preserve
 definitional equality in certain cases. -/
-class algebra (R : Type u) (A : Type v) [comm_ring R] [ring A] extends has_scalar R A :=
-(to_fun : R → A) [hom : is_ring_hom to_fun]
+@[nolint has_inhabited_instance]
+class algebra (R : Type u) (A : Type v) [comm_semiring R] [semiring A] extends has_scalar R A :=
+(to_fun : R →+* A)
 (commutes' : ∀ r x, x * to_fun r = to_fun r * x)
 (smul_def' : ∀ r x, r • x = to_fun r * x)
 end prio
 
-def algebra_map {R : Type u} (A : Type v) [comm_ring R] [ring A] [algebra R A] (x : R) : A :=
-algebra.to_fun A x
+/-- Embedding `R →+* A` given by `algebra` structure. -/
+def algebra_map (R : Type u) (A : Type v) [comm_semiring R] [semiring A] [algebra R A] : R →+* A :=
+algebra.to_fun R A
 
 namespace algebra
 
 variables {R : Type u} {S : Type v} {A : Type w}
-variables [comm_ring R] [comm_ring S] [ring A] [algebra R A]
 
-instance : is_ring_hom (algebra_map A : R → A) := algebra.hom _ A
+section semiring
 
-variables (A)
-@[simp] lemma map_add (r s : R) : algebra_map A (r + s) = algebra_map A r + algebra_map A s :=
-is_ring_hom.map_add _
-
-@[simp] lemma map_neg (r : R) : algebra_map A (-r) = -algebra_map A r :=
-is_ring_hom.map_neg _
-
-@[simp] lemma map_sub (r s : R) : algebra_map A (r - s) = algebra_map A r - algebra_map A s :=
-is_ring_hom.map_sub _
-
-@[simp] lemma map_mul (r s : R) : algebra_map A (r * s) = algebra_map A r * algebra_map A s :=
-is_ring_hom.map_mul _
-
-variables (R)
-@[simp] lemma map_zero : algebra_map A (0 : R) = 0 :=
-is_ring_hom.map_zero _
-
-@[simp] lemma map_one : algebra_map A (1 : R) = 1 :=
-is_ring_hom.map_one _
-variables {R A}
+variables [comm_semiring R] [comm_semiring S] [semiring A] [algebra R A]
 
 /-- Creating an algebra from a morphism in CRing. -/
-def of_ring_hom (i : R → S) (hom : is_ring_hom i) : algebra R S :=
+def of_ring_hom (i : R →+* S) : algebra R S :=
```

Comment:
Would it make sense to (also) have `ring_hom.as_algebra` to enable dot notation?